config/custom.go

```diff
@@ -220,53 +285,66 @@ func (urls WeightedURLs) Flatten() ([]string, []uint32, error) {
 	return str, wghts, nil
 }
 
+// DecodePolicyBase64Hook creates a mapstructure DecodeHookFunc.
 func DecodePolicyBase64Hook() mapstructure.DecodeHookFunc {
 	return func(f, t reflect.Type, data interface{}) (interface{}, error) {
-		if t != reflect.TypeOf([]Policy{}) {
-			return data, nil
-		}
+		switch t {
+		case reflect.TypeOf([]Policy{}):
+			str, ok := data.([]string)
+			if !ok {
+				return data, nil
+			}
 
-		str, ok := data.([]string)
-		if !ok {
-			return data, nil
-		}
+			if len(str) != 1 {
+				return nil, fmt.Errorf("base64 policy data: expecting 1, got %d", len(str))
+			}
 
-		if len(str) != 1 {
-			return nil, fmt.Errorf("base64 policy data: expecting 1, got %d", len(str))
-		}
+			bs, err := base64.StdEncoding.DecodeString(str[0])
+			if err != nil {
+				return nil, fmt.Errorf("base64 decoding policy data: %w", err)
+			}
 
-		bytes, err := base64.StdEncoding.DecodeString(str[0])
-		if err != nil {
-			return nil, fmt.Errorf("base64 decoding policy data: %w", err)
-		}
+			out := []map[interface{}]interface{}{}
+			if err = yaml.Unmarshal(bs, &out); err != nil {
+				return nil, fmt.Errorf("parsing base64-encoded policy data as yaml: %w", err)
+			}
 
-		out := []map[interface{}]interface{}{}
-		if err = yaml.Unmarshal(bytes, &out); err != nil {
-			return nil, fmt.Errorf("parsing base64-encoded policy data as yaml: %w", err)
+			return out, nil
 		}
-
-		return out, nil
+		return data, nil
 	}
 }
 
+// DecodePolicyHookFunc creates a mapstructure DecodeHookFunc.
 func DecodePolicyHookFunc() mapstructure.DecodeHookFunc {
 	return func(f, t reflect.Type, data interface{}) (interface{}, error) {
-		if t != reflect.TypeOf(Policy{}) {
-			return data, nil
-		}
+		switch t {
```

Comment:
my understanding of how mapstructure hooks work is that you create a separate hook for each type, instead of making a large single function which tries to handle multiple ones. 

also then you could test them individually.